cmd/harmony/config/config.go

```diff
@@ -255,67 +255,102 @@ func parseNetworkType(nt string) nodeconfig.NetworkType {
 	}
 }
 
-func getDefaultSyncConfig(nt nodeconfig.NetworkType) syncConfig {
+func GetDefaultSyncConfig(nt nodeconfig.NetworkType) SyncConfig {
 	switch nt {
 	case nodeconfig.Mainnet:
-		return defaultMainnetSyncConfig
+		return DefaultMainnetSyncConfig
 	case nodeconfig.Testnet:
-		return defaultTestNetSyncConfig
+		return DefaultTestNetSyncConfig
 	case nodeconfig.Localnet:
 		return defaultLocalNetSyncConfig
 	default:
 		return defaultElseSyncConfig
 	}
 }
 
-var dumpConfigCmd = &cobra.Command{
-	Use:   "dumpconfig [config_file]",
-	Short: "dump the config file for harmony binary configurations",
-	Long:  "dump the config file for harmony binary configurations",
-	Args:  cobra.MinimumNArgs(1),
-	Run: func(cmd *cobra.Command, args []string) {
-		nt := getNetworkType(cmd)
-		config := getDefaultHmyConfigCopy(nt)
-
-		if err := writeHarmonyConfigToFile(config, args[0]); err != nil {
-			fmt.Println(err)
-			os.Exit(128)
-		}
-	},
-}
-
-func registerDumpConfigFlags() error {
-	return cli.RegisterFlags(dumpConfigCmd, []cli.Flag{networkTypeFlag})
+func GetDefaultDNSSyncConfig(nt nodeconfig.NetworkType) DNSSyncConfig {
+	config := DNSSyncConfig{
+		LegacySyncing: false,
+		DNSZone:       nodeconfig.GetDefaultDNSZone(nt),
+		DNSPort:       nodeconfig.GetDefaultDNSPort(nt),
+		LegacyServer:  true,
+		LegacyClient:  true,
+	}
+	switch nt {
+	case nodeconfig.Mainnet:
+	case nodeconfig.Testnet:
+	case nodeconfig.Localnet:
+	default:
+		config.LegacyClient = false
+	}
+	return config
 }
 
-func loadHarmonyConfig(file string) (harmonyConfig, error) {
+func LoadHarmonyConfig(file string) (HarmonyConfig, error) {
 	b, err := ioutil.ReadFile(file)
 	if err != nil {
-		return harmonyConfig{}, err
+		return HarmonyConfig{}, err
 	}
-
-	var config harmonyConfig
-	if err := toml.Unmarshal(b, &config); err != nil {
-		return harmonyConfig{}, err
+	var header struct {
+		Version string
+	}
+	if err := toml.Unmarshal(b, &header); err != nil {
+		return HarmonyConfig{}, fmt.Errorf("reading config version: %w", err)
+	}
+	var config HarmonyConfig
+	// lexicographicaly compare
+	if header.Version <= v1TOMLConfigVersion {
+		config, err = loadHarmonyConfigV1(b)
```

Comment:
It would be better to `loadHarmonyConfigV1` returns `harmonyConfigV1` and then parse V1 to V2 explicitly here. Which will increase the readability as well as give convenience for future updates.

Reply:
My idea is that we should have a file per old config version that we want to support, e.g. v1.go, v2.go etc, which should contain the old struct together with the code to convert that struct into the latest HarmonyConfig. that way we do not clutter main load function with every possible converting logic, but rather have that logic contained close to its struct.
I am not sure that having this convert logic for each version in the one place will increase readability, considering that we might have a number of versions in the future.
Or maybe I am missing the point, then correct me, please.

Reply:
I am very glad to say that I definitely agree every point you have in the comment. What I have in mind is that, currently `LoadHarmonyConfig` is actually doing two jobs:

1. Load the config from toml file to its own version.
2. Parse the config to the latest version.

So I would love to see this two jobs are defined in two functions:

1. In each version file, load `toml` file to its own version. This will also handle minor version updates, E.g. v1.0.4 to v1.0.5.
2. Create a file `parse.go` to parse the legacy version to the latest.
```
// v1.go
func loadHarmonyConfigV1(file string) harmonyConfigV1 {}

// v2.go
func loadHarmonyConfigV2(file string) harmonyConfigV2 {}

// parse.go
func parseConfigToLatest(version vesion.Version, config interface{}) HarmonyConfig {
    switch version {
    ...
    }
}

func parseConfigV1ToV2(v1 harmonyConfigV1) harmonyConfigV2 {...}
func parseConfigV2ToV3(v2 harmonyConfigV2) harmonyConfigV3 {...}
```

In this case, when a new version of harmonyConfig is launched, there will be minimum update in the legacy file and functions.